state/cloudimagemetadata/image_test.go

```diff
@@ -0,0 +1,307 @@
+	added := cloudimagemetadata.Metadata{
+		Storage:     storage,
+		VirtType:    virtType,
+		Arch:        arch,
+		Series:      series,
+		RegionAlias: regionAlias,
+		RegionName:  regionName,
+		Endpoint:    endpoint,
+		Stream:      stream,
+	}
+	err := s.storage.AddMetadata(added)
+	c.Assert(err, jc.ErrorIsNil)
+
+	s.assertMetadata(c, added)
+}
+
+func (s *cloudImageMetadataSuite) TestAllMetadata(c *gc.C) {
+	metadata, err := s.storage.AllMetadata()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(metadata, gc.HasLen, 0)
+
+	m := cloudimagemetadata.Metadata{
+		Series:      "quantal",
+		Arch:        "amd64",
+		Stream:      "test",
+		Storage:     "storage",
+		VirtType:    "virtType",
+		RegionAlias: "regionAlias",
+		RegionName:  "regionName",
+		Endpoint:    "endpoint",
+	}
+	s.addMetadataDoc(c,
+		m.Series,
+		m.Arch,
+		m.Stream,
+		m.Storage,
+		m.VirtType,
+		m.RegionAlias,
+		m.RegionName,
+		m.Endpoint,
+	)
+
+	metadata, err = s.storage.AllMetadata()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(metadata, gc.HasLen, 1)
+	expected := []cloudimagemetadata.Metadata{m}
+	c.Assert(metadata, jc.SameContents, expected)
+
+	m.Arch = "my one"
+	s.addMetadataDoc(c,
+		m.Series,
+		m.Arch,
+		m.Stream,
+		m.Storage,
+		m.VirtType,
+		m.RegionAlias,
+		m.RegionName,
+		m.Endpoint,
+	)
+
+	metadata, err = s.storage.AllMetadata()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(metadata, gc.HasLen, 2)
+	expected = append(expected, m)
+	c.Assert(metadata, jc.SameContents, expected)
+}
+
+func (s *cloudImageMetadataSuite) TestFindMetadata(c *gc.C) {
+	m := cloudimagemetadata.Metadata{
+		Series:      "quantal",
+		Arch:        "amd64",
+		Stream:      "test",
+		Storage:     "storage",
+		VirtType:    "virtType",
+		RegionAlias: "regionAlias",
+		RegionName:  "regionName",
+		Endpoint:    "endpoint",
+	}
+
+	_, err := s.storage.FindMetadata(m.Series, m.Arch, m.Stream)
+	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+
+	s.addMetadataDoc(c,
+		m.Series,
+		m.Arch,
+		m.Stream,
+		m.Storage,
+		m.VirtType,
+		m.RegionAlias,
+		m.RegionName,
+		m.Endpoint,
+	)
+	s.assertMetadata(c, m)
+}
+
+func (s *cloudImageMetadataSuite) TestAddMetadataDuplicate(c *gc.C) {
+	metadata := cloudimagemetadata.Metadata{
+		Series: "quantal",
+		Arch:   "amd64",
+		Stream: "test",
+	}
+	for i := 0; i < 2; i++ {
+		err := s.storage.AddMetadata(metadata)
+		c.Assert(err, jc.ErrorIsNil)
+		s.assertMetadata(c, metadata)
+	}
+	all, err := s.storage.AllMetadata()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(all, gc.HasLen, 1)
+	expected := []cloudimagemetadata.Metadata{metadata}
+	c.Assert(all, jc.SameContents, expected)
+
+}
+
+func (s *cloudImageMetadataSuite) TestAddMetadataConcurrent(c *gc.C) {
+	metadata0 := cloudimagemetadata.Metadata{
+		Series: "quantal",
+		Arch:   "amd64",
+		Stream: "test",
+	}
+	metadata1 := cloudimagemetadata.Metadata{
+		Series: "quantal",
+		Arch:   "amd64",
+		Stream: "test2",
```

Comment:
i think it would be better to test concurrent writes with records that will actually exercise the insert vs update logic in the buildTxn loop.